packages/gensx-openai/src/composition.tsx

```diff
@@ -0,0 +1,359 @@
+import { gsx } from "gensx";
+import {
+  ChatCompletion as ChatCompletionOutput,
+  ChatCompletionChunk,
+  ChatCompletionCreateParamsNonStreaming,
+  ChatCompletionCreateParamsStreaming,
+  ChatCompletionMessageParam,
+} from "openai/resources/chat/completions";
+import { Stream } from "openai/streaming";
+import { z } from "zod";
+
+import { GSXStructuredOutput, GSXTool, OpenAIContext } from "./index.js";
+
+// Base types for raw OpenAI chat completion
+type RawCompletionProps = Omit<
+  ChatCompletionCreateParamsNonStreaming,
+  "stream" | "tools" | "response_format"
+>;
+
+type RawCompletionReturn = ChatCompletionOutput;
+
+// Raw completion component that directly calls OpenAI
+export const RawCompletion = gsx.Component<
+  RawCompletionProps,
+  RawCompletionReturn
+>("RawCompletion", async (props) => {
+  const context = gsx.useContext(OpenAIContext);
+  if (!context.client) {
+    throw new Error(
+      "OpenAI client not found in context. Please wrap your component with OpenAIProvider.",
+    );
+  }
+
+  return context.client.chat.completions.create({
+    ...props,
+    stream: false,
+  } as ChatCompletionCreateParamsNonStreaming);
+});
+
+// Stream transform component
+type StreamTransformProps = RawCompletionProps & {
+  stream: true;
+  tools?: GSXTool<z.ZodObject<z.ZodRawShape>>[];
+};
+
+type StreamTransformReturn = Stream<ChatCompletionChunk>;
+
+export const StreamTransform = gsx.Component<
+  StreamTransformProps,
+  StreamTransformReturn
+>("StreamTransform", async (props) => {
+  const context = gsx.useContext(OpenAIContext);
+  if (!context.client) {
+    throw new Error(
+      "OpenAI client not found in context. Please wrap your component with OpenAIProvider.",
+    );
+  }
+
+  const { stream, tools, ...rest } = props;
+  const openAITools = tools?.map((tool) => tool.toOpenAITool());
+
+  return context.client.chat.completions.create({
+    ...rest,
+    tools: openAITools,
+    stream: true,
+  } as ChatCompletionCreateParamsStreaming);
+});
+
+// Tool execution component
+interface ToolExecutorProps {
+  tools: GSXTool<z.ZodObject<z.ZodRawShape>>[];
+  toolCalls: NonNullable<
+    ChatCompletionOutput["choices"][0]["message"]["tool_calls"]
+  >;
+  messages: ChatCompletionMessageParam[];
+}
+
+type ToolExecutorReturn = ChatCompletionOutput;
+
+export const ToolExecutor = gsx.Component<
```

Comment:
I wonder if naming should be nouns or verbs? 

I kind of like verbs for things like this. RunTool perhaps?